### DISCUSSION TITLE
API proposals

### DISCUSSION THREAD

**Discussion by vsysoev** [IntegraSDL/pygats#3] [Ideas]
All,
Need to think about user API for the library to provide simple and friendly interface to write tests and do research.
My thoughts on this:
Two possible ways must be researched:
1. Functional approach to API
2. Chaining approach to API use Object-Oriented Approach

## Functional approach to API

1. Try to be as much as possible in functional programming approach https://docs.python.org/3/howto/functional.html
3. Keep functions as pure as possible.
4. Use first-class function
5. Keep all formatters(for ex. markdown, text, docx, html) as separate entity in library and pass formatter functions as parameter.
6. Try to avoid global states

## Chaining approach to API

Might be Object-Oriented

for ex:
```
t.step1().step2().step3().step4().passed()
```

We need to discuss pros and cons of these two approaches. Please share me your thoughts on this

## Functions grouping

All functions in API might be grouped by functional area:
* output and input functions;
* processing functions (image/state or something else);
* keyboard/mouse control function

**Comment by chessplayer123** [comment]
OOP can become the main paradigm of our project.

Obvious benefits include:

1. Application state can be described by a class (previous state, possible transitions, etc.)
3. Working with a screenshot can be wrapped in a class (search for text, outlines, etc.)
4. State management can be done through class methods (press a key, button, etc.)
5. Working with output data can be abstracted into a separate class.
6. Development can be parallelized by several people. Each will provide a class interface (a set of methods, what they accept, what they return) and each, doing his part, will be able to rely on other classes.

_This message expresses the general opinion of the team._

**Reply by vsysoev** [reply]
Good points! but "obvious" is not a good argument. We have to architect library API in this discussion. So we need to 
Let me express my opinion on your points:
1. pymagic-autogui is a library not an application. It supposed to provide bridge between pyautogui, tesseract, opencv and ai staff. I think state should be external entity. IMHO external application or test system should manage state.
2. Are we going to provide screenshot as entity of API?
3. Can be done in any way but no arguments why it supposed to be done this way
4. Don't see any arguments why it should be done this way
5. Low coupling in API can help with this not only OOP approach.

Let's meet to discuss this.
